topdown/trusted_identity_providers.go

```diff
@@ -0,0 +1,121 @@
+package topdown
+
+import (
+	"errors"
+	"fmt"
+	"github.com/coreos/go-oidc"
+	jwt "github.com/square/go-jose/v3/jwt"
+	"golang.org/x/net/context"
+	"sync"
+	"time"
+)
+
+
+var globalTrustedIdProviderManager TrustedIdProviderManagerImpl
+var globalTrustedIssuerContextTimeout = time.Second * time.Duration(5)
+
+// May be used to manage an token verification against an entire collection
+// of trusted IdP's.
+type TrustedIdProviderManager interface {
+
+	// On nil error, IDToken contains the parsed jwt token iff there exist
+	// at least one trusted identity provider in this managed set of IdP's
+	// that successfully validated this token. Else, (if not valid or not
+	// trusted or an error is encountered), a non-nil error returned and
+	// IDToken is nil.
+	VerifyToken(token *string) (*oidc.IDToken, error)
+}
+
+// Private struct to implement interface, TrustedIdProviderManager.
+// We use a sync map to safely manage a collection of trusted
+// issuers and their verifiers.
+type TrustedIdProviderManagerImpl struct {
+
+	// Thread safe Map of string of trustedIssuer to *oidc.IDTokenVerifier trustedIssuerVerifier.
+	trustedVerifiers sync.Map
+
+}
+
+func CreateOrGetVerifier(idp *string) (*oidc.IDTokenVerifier, error) {
+
+	// If we already have a verifier for this issuer, use it.
+	if loadedVerifier, ok := globalTrustedIdProviderManager.trustedVerifiers.Load(*idp); ok {
```

Comment:
Do we ever need to be worried about the global cache of these becoming stale or needing to be invalidated?

Reply:
Great question, though they'll never be stale or invalid. 

The state being saved in those objects are the public keys of the IdP's (we only want to request the IdP's keys 1x, for repeated verifications). I double-checked the RFC for token exchange just to be sure, but it appears that the public keys never expire.  Manually looking at a few example keys, no TTL's or EXP fields seem to corroborate this:
(our test idp)
https://dev-530681.oktapreview.com/oauth2/v1/keys
(google's prod keys)
https://www.googleapis.com/oauth2/v3/certs